File: EPI/10.10/swift/10.10.swift
public class Node<T: Hashable> {
    public let value: T
    public let left: Node?
    public let right: Node?

    public convenience init?(inorder: [T], preorder: [T]) {
        var inorderIndexes = [T: Int]()
        for (index, value) in inorder.enumerate() {
            inorderIndexes[value] = index
        }

        self.init(
            inorder: inorder[0..<inorder.count],
            preorder: preorder[0..<preorder.count],
            inorderIndexes: inorderIndexes
        )
    }

    init?(inorder: ArraySlice<T>, preorder: ArraySlice<T>, inorderIndexes: [T: Int]) {
        if inorder.isEmpty {
            return nil
        }

        self.value = preorder.first!

        let inorderRootIndex = inorderIndexes[self.value]!
        let leftCount = inorderRootIndex - inorder.startIndex
        let preorderRightStartIndex = preorder.startIndex + 1 + leftCount

        self.left = Node(
            inorder: inorder[inorder.startIndex..<inorderRootIndex],
            preorder: preorder[preorder.startIndex + 1..<preorderRightStartIndex],
            inorderIndexes: inorderIndexes
        )

        self.right = Node(
            inorder: inorder[inorderRootIndex + 1..<inorder.endIndex],
            preorder: preorder[preorderRightStartIndex..<preorder.endIndex],
            inorderIndexes: inorderIndexes
        )
    }
}

File: EPI/10.10/typescript/10.10.ts
export interface Node<T> {
    value: T;
    left?: Node<T>;
    right?: Node<T>;
}

export function reconstruct<T>(inOrder: T[], preOrder: T[]): Node<T> | undefined {
    return reconstructRec(inOrder, 0, preOrder, 0).root;
}

function reconstructRec<T>(inOrder: T[], i: number, preOrder: T[], p: number,
    nextInOrderAncestorValue?: T): {root?: Node<T>, size: number} {

    if (i >= inOrder.length || p >= preOrder.length) {
        return {size: 0};
    }

    const rootValue = preOrder[p];

    // Move past the root.
    p += 1;

    const left = inOrder[i] !== rootValue ?
        reconstructRec(inOrder, i, preOrder, p, rootValue) :
        {size: 0};

    // Move past the left subtree.
    i += left.size + 1;
    p += left.size;

    const right = i < inOrder.length && inOrder[i] !== nextInOrderAncestorValue ?
        reconstructRec(inOrder, i, preOrder, p, nextInOrderAncestorValue) :
        {size: 0};

    return {
        root: {value: rootValue, left: left.root, right: right.root},
        size: 1 + left.size + right.size
    };
}
